hedera-node/test-clients/README.md

```diff
@@ -1,11 +1,250 @@
+but this requires single-threaded execution and exclusive use of Ed25519 keys, since the ECDSA signing process has
+inherent randomness.
+
+Either way, we can now directly access and manipulate the network's internal state; and we can submit transactions
+that would be rejected by a real network. Furthermore, in repeatable mode, `SpecOperation`'s that would normally
+sleep in real time switch to immediately advancing virtual time the requested amount. This lets us test staking
+rewards, contract expiration, and other time-dependent behaviors orders of magnitude faster than we could otherwise.
+
+<img src="./docs/assets/embedded-network.png" width="900">
+
+## Style guide
+
+As with any test, a `@HapiTest` should be clear, concise, and focused. The main challenge is when the `SpecOperation`
+DSL does not make it easy to isolate the behavior being tested. For example, say we want to validate that the fee
+for an auto-account creation triggered by a `ScheduleSign` transaction is charged to the payer account of the
+scheduled `CryptoTransfer` and not the `ScheduleSign` payer. That intent could easily be obscured by the boilerplate
+needed to create the schedule, sign it, get the records involved, and perhaps even check the balances of the payer
+accounts involved before and after the `ScheduleSign`.
+
+There are many ways to manage such complexity. We could,
+* Extend the DSL, especially the object-oriented DSL in the `c.h.s.bdd.spec.dsl` package, to encapsulate the
+   boilerplate and expose only the intent.
+* Extract the setup into a `@BeforeAll` method, using an injected `SpecManager` to create prerequisite entities
+   needed by this and other `@HapiTest` methods in the test class.
+* Write a helper method that returns a grouped `SpecOperation` to encapsulate the setup operations.
+* Accept a slightly more verbose test method, but add rigorous comments to explain each step.
+
+There is no "best" approach. For example, the effort of extending the DSL is probably only justified when the new
+constructs will apply to many tests. But it is essential to achieve a final result that meets two criteria:
+1. Any qualified PR reviewer can understand the behavior being asserted **in a single reading**; and,
+2. The `@HapiTest` javadoc links any context necessary to understand the motivation for the test.
+
+### The `@HapiTest` checklist
+To help consistently meet these criteria, we use the following minimum checklist for every `@HapiTest`:
+- [x] The `@DisplayName`, when combined with the display names of any enclosing test classes, forms a single, clear
+  phrase that fully summarizes the behavior being tested.
+- [x] If the test specifies behavior from a HIP, its javadoc links to the section in the markdown file in the
+  [`hedera-improvement-proposal` repo](https://github.com/hashgraph/hedera-improvement-proposal) outlining the behavior.
+- [x] The test includes no more than seven (7) `SpecOperation`'s.
```

Comment:
Q: is this a best practices recommendation or a limitation?
Also does this include setup operations or do all SpecOperation occur after setup transactions?
Some intricate flows might get close or require more operations

Reply:
I consider it a best practice/requirement...you can see [here](https://github.com/hashgraph/hedera-services/pull/14105/files#diff-ebfd622a28cce1b2e4e1bce2078d4c8ead917f05384552f2b47c801c5a20f143R64) I just added probably a quite involved test class that goes through two software upgrades while removing and adding nodes and validating staking; but no single `@HapiTest` uses more than 6 operations.